src/transformers/generation/utils.py

```diff
@@ -3404,184 +3439,241 @@ def _beam_search(
                 model_kwargs["encoder_outputs"].get("hidden_states") if output_hidden_states else None
             )
 
+        # 3. init running tensors
+        # per batch, beam-item holding current token in loop and completed sequences
+        running_sequences = unflatten_beam_dim(input_ids, batch_size, num_beams)
+        sequences = running_sequences.clone().detach()
+        # per batch, beam-item score, logprobs
         # initialise score of first beam with 0 and the rest with -1e9. This makes sure that only tokens
         # of the first beam are considered to avoid sampling the exact same tokens across all beams.
-        beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
-        beam_scores[:, 1:] = -1e9
-        beam_scores = beam_scores.view((batch_size * num_beams,))
+        running_beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
+        running_beam_scores[:, 1:] = -1e9
+        beam_scores = torch.ones((batch_size, num_beams), dtype=torch.float, device=input_ids.device) * -1e9
+        # per batch, beam-item state bit indicating if sentence has finished.
+        is_sent_finished = torch.zeros((batch_size, num_beams), dtype=torch.bool, device=input_ids.device)
+        # per batch, beam-item state bit indicating if there are valid continuations.
+        next_token_hits_stopping_criteria = torch.zeros(
+            (batch_size, num_beams), dtype=torch.bool, device=input_ids.device
+        )
+        # per batch beam indices
+        running_beam_indices = torch.ones((batch_size, num_beams, 0), dtype=torch.int32, device=input_ids.device)
+        beam_indices = running_beam_indices.clone().detach()
+
+        # 4. define custom stopping criteria (depends on the state of the beams and their scores)
+        def beam_search_has_unfinished_sequences(
```

Comment:
to put outside the context of this function otherwise impossible to change / customize and just not waht we usually do